src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs

```diff
@@ -280,118 +319,253 @@ private string GenerateForCollection(TypeMetadata typeMetadata)
                     _ => throw new NotSupportedException()
                 };
 
                 string metadataInitSource = @$"_{typeFriendlyName} = {collectionTypeInfoValue};";
-                return GenerateForType(typeMetadata, metadataInitSource);
+
+                return GenerateForType(typeGenerationSpec, metadataInitSource, serializeFuncSource);
             }
 
-            private string GenerateForObject(TypeMetadata typeMetadata)
+            private string GenerateFastPathFuncForEnumerable(string typeInfoRef, string serializeMethodName, bool canBeNull, bool isArray, TypeGenerationSpec valueTypeGenerationSpec)
             {
-                string typeCompilableName = typeMetadata.CompilableName;
-                string typeFriendlyName = typeMetadata.FriendlyName;
+                string? writerMethodToCall = GetWriterSerializationMethod(valueTypeGenerationSpec.Type);
+                string valueToWrite = $"{ValueVarName}[i]";
+                string lengthPropName = isArray ? "Length" : "Count";
+
+                string serializationLogic;
+                if (writerMethodToCall != null)
+                {
+                    serializationLogic = $"{writerMethodToCall}Value({valueToWrite});";
+                }
+                else
+                {
+                    serializationLogic = GetSerializeLogicForNonPrimitiveType(valueTypeGenerationSpec.TypeInfoPropertyName, valueToWrite, valueTypeGenerationSpec.GenerateSerializationLogic);
+                }
+
+                string fastPathLogic = $@"{WriterVarName}.WriteStartArray();
+
+        for (int i = 0; i < {ValueVarName}.{lengthPropName}; i++)
+        {{
+            {serializationLogic}
+        }}
+
+        {WriterVarName}.WriteEndArray();";
+
+                return GenerateFastPathFuncForType(serializeMethodName, typeInfoRef, fastPathLogic, canBeNull);
+            }
+
+            private string GenerateFastPathFuncForDictionary(
+                string typeInfoRef,
+                string serializeMethodName,
+                bool canBeNull,
+                TypeGenerationSpec keyTypeGenerationSpec,
+                TypeGenerationSpec valueTypeGenerationSpec)
+            {
+                string? writerMethodToCall = GetWriterSerializationMethod(valueTypeGenerationSpec.Type);
+                const string pairVarName = "pair";
+                string keyToWrite = $"{pairVarName}.Key";
+                string valueToWrite = $"{pairVarName}.Value";
+
+                string serializationLogic;
+                if (writerMethodToCall != null)
+                {
+                    serializationLogic = $"{writerMethodToCall}({keyToWrite}, {valueToWrite})";
```

Comment:
This is missing a `;` at the end, which results in a compiler error if this code path is taken.

It also indicates we probably don't have a test case for this path.

Reply:
Thanks. Fixed and I added a test.